github.go

```diff
@@ -0,0 +1,267 @@
+	"golang.org/x/oauth2"
+)
+
+var errFileNotFound = fmt.Errorf("file not found on github")
+
+// GitHub is a way to store checkup results in a GitHub repository.
+type GitHub struct {
+	AccessToken     string `json:"access_token"`
+	RepositoryOwner string `json:"repository_owner"`
+	RepositoryName  string `json:"repository_name"`
+	CommitterName   string `json:"committer_name"`
+	CommitterEmail  string `json:"committer_email"`
+	Branch          string `json:"branch"`
+	Dir             string `json:"dir"`
+
+	// Check files older than CheckExpiry will be
+	// deleted on calls to Maintain(). If this is
+	// the zero value, no old check files will be
+	// deleted.
+	CheckExpiry time.Duration `json:"check_expiry,omitempty"`
+
+	client *github.Client
+}
+
+func (gh *GitHub) ensureClient() error {
+	if gh.client != nil {
+		return nil
+	}
+
+	if gh.AccessToken == "" {
+		return fmt.Errorf("Please specify access_token in storage configuration")
+	}
+
+	gh.client = github.NewClient(oauth2.NewClient(
+		oauth2.NoContext,
+		oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: gh.AccessToken},
+		),
+	))
+
+	return nil
+}
+
+func (gh GitHub) fullPathName(filename string) string {
+	if strings.HasPrefix(filename, gh.Dir) {
+		return filename
+	} else {
+		return filepath.Join(gh.Dir, filename)
+	}
+}
+
+func (gh GitHub) readFile(filename string) ([]byte, string, error) {
+	if err := gh.ensureClient(); err != nil {
+		return nil, "", err
+	}
+
+	contents, _, resp, err := gh.client.Repositories.GetContents(
+		context.Background(),
+		gh.RepositoryOwner,
+		gh.RepositoryName,
+		gh.fullPathName(filename),
+		&github.RepositoryContentGetOptions{Ref: "heads/" + gh.Branch},
+	)
+	if err != nil {
+		if resp.StatusCode == 404 {
+			return nil, "", errFileNotFound
+		}
+		return nil, "", err
+	}
+
+	decoded, err := contents.GetContent()
+	return []byte(decoded), *contents.SHA, err
+}
+
+func (gh GitHub) writeFile(filename string, sha string, contents []byte) error {
+	if err := gh.ensureClient(); err != nil {
+		return err
+	}
+
+	var err error
+	var writeFunc func(context.Context, string, string, string, *github.RepositoryContentFileOptions) (*github.RepositoryContentResponse, *github.Response, error)
+	opts := &github.RepositoryContentFileOptions{
+		Message: github.String(fmt.Sprintf("[checkup] store %s [ci skip]", filename)),
+		Content: contents,
+		Committer: &github.CommitAuthor{
+			Name:  &gh.CommitterName,
+			Email: &gh.CommitterEmail,
+		},
+	}
+
+	if gh.Branch != "" {
+		opts.Branch = &gh.Branch
+	}
+
+	// If no SHA specified, then create the file.
+	// Otherwise, update the file at the specified SHA.
+	if sha == "" {
+		writeFunc = gh.client.Repositories.CreateFile
+		fmt.Printf("creating %s on branch=%s\n", gh.fullPathName(filename), gh.Branch)
+	} else {
+		opts.SHA = github.String(sha)
+		writeFunc = gh.client.Repositories.UpdateFile
+		fmt.Printf("updating %s on branch=%s\n", gh.fullPathName(filename), gh.Branch)
+	}
+
+	_, _, err = writeFunc(
```

Comment:
Yep! They're `*github.RepositoryContentResponse` and `*github.Response`. [CreateFile](https://godoc.org/github.com/google/go-github/github#RepositoriesService.CreateFile) and [UpdateFile](https://godoc.org/github.com/google/go-github/github#RepositoriesService.UpdateFile) have the same function signature.